File: practice2/solution.py
def formula_simps(n: int, a: int, b: int, f) -> int:

    h = (b - a) / float(n)
    result = f(a) + f(b)

    for i in range(n):
        x: float = a + i * h
        if i % 2 == 0:
            result += 2 * f(x)
        else:
            result += 4 * f(x)
    return result * h / 3


def y1(x): return x ** 3 - 12 * x ** 2 + 4
def y2(x): return -x ** 3 + x ** 2 - 4


def area(a, b, n, f1=y1, f2=y2): return abs(
    abs(formula_simps(n, a, b, f1)) - abs(formula_simps(n, a, b, f2)))


if __name__ == "__main__":
    a, b, = 3, 6
    area = area(y1, y2, a, b, 10**10)
    print(f"Приближенное значение: {area}")

File: practice2/solution.cpp
#include "solution.h"
#include <cmath>
#include <iostream>

double simpson_rule(const std::function<double(double)> &f, double a, double b, int n)
{
    if (n % 2 == 1)
        n++;

    double h = (b - a) / n;
    double sum = f(a) + f(b);

    for (int i = 1; i < n; i += 2)
    {
        double x = a + i * h;
        sum += 4.0 * f(x);
    }

    for (int i = 2; i < n - 1; i += 2)
    {
        double x = a + i * h;
        sum += 2.0 * f(x);
    }

    return (h / 3.0) * sum;
}

double func_y1(double x)
{
    return std::pow(x, 3) - 12.0 * std::pow(x, 2) + 4.0;
}

double func_y2(double x)
{
    return -std::pow(x, 3) + std::pow(x, 2) - 4.0;
}

double diff(double x)
{
    return func_y2(x) - func_y1(x);
}

double area(double a, double b, int n)
{
    return std::abs(simpson_rule(diff, a, b, n));
}